# irscalc

This repository contains analysis scripts for:
Ishigami et al. (2021) A single m6A modification in U6 snRNA diversifies exon sequence at the 5â€™ splice site. (Nature Communications)

Environment:
python 3.8.5 with packages pysam, Biopython and numpy installed

Script gtfbamcounts.py counts the reads mapped on splice junctions and classify them depending on their splicing pattern.
Usage: python3 gtfbamcount.py [BAM file name] [GTF file name] [output bamcount file name]

Script irsseq.py calculates the IRS/PCS scores of each intron based on the count file generated by gtfbamcounts.py and export them with their sequence based on genomic DNA sequence FASTA.
Usage: python3 gtfbamcount.py [bamcount file list: control] [bamcount file list: treated] [genome FASTA file name] [output file name]
Bamcount filenames are delimited by commas without spaces.
Example: python3 gtfbamcount.py wt1.bam.txt,wt2.bam.txt,wt3.bam.txt ko1.bam.txt,ko2.bam.txt,ko3.bam.txt genome.fa outfile.txt
